### PR TITLE
Replace Unit with wildcards in method signatures

### LINKED ISSUE #381: Improve performance through polymorphism
Right now, many methods accept `IO[E, Unit]`.

For example:

```scala
IO.bracket / io.bracket
IO.bracket0 / io.bracket0
io.fork0
io.ensuring
io.bracketOnError
io.managed
io.onError
io.onInterrupt
io.onTermination
IO.supervised / io.supervised
Promise.bracket
```

This means that if you pass an `IO[E, A]`, you must first call `void` to discard the return value.

On the one hand, this forces a user to be explicit: they must know the return value will be discarded. This makes it harder to get order of parameters wrong (e.g. `io.bracket`). On the other hand, calling `void` involves an extraneous `flatMap` operation that has impacts on performance critical code.

To improve performance, we can relax the constraint that `Unit` be returned from many (perhaps all) of these methods, and instead accept `IO[E, Any]`. 

This should not involve adding any new functionality, just changing some type signatures. After they are changed, some `.void` calls can be eliminated inside ZIO to improve performance.

### PR DESCRIPTION
**Pull request by mijicd**
This PR resolves #381.

### CODE REVIEW

**Comment by jdegoes** [comment]
@mijicd Overall, this looks great! However, we decided in that ticket to use wildcards `_` instead of `Any`, to better communicate the intent in the API. Should be a minor tweak, though.

**Comment by mijicd** [comment]
@jdegoes done. However, I couldn't overcome all `Any` occurrences, i.e. the `IO.unit` used in folds as empty case are still ascribed as `IO[Nothing, Any]`.

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 1]
```diff
@@ -339,7 +339,7 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * because it's possible the fiber will be interrupted after acquisition but
    * before the finalizer is added.
    */
-  final def ensuring(finalizer: IO[Nothing, Any]): IO[E, A] =
+  final def ensuring(finalizer: IO[Nothing, _]): IO[E, A] =
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: With all these changes, you can look for expressions like `io.void` and probably simplify to `io` most of the time (perhaps changing a `*>` into `<*` too)

Comment (mijicd) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: Good point, will take care of that.

File: core/shared/src/main/scala/scalaz/zio/Promise.scala [hunk 2]
```diff
@@ -183,7 +183,7 @@ object Promise {
                   case (p, io) => io.flatMap(c => pRef.set(Some((c, p))) *> IO.now(p))
                 }.uninterruptibly
             b <- p.get
-          } yield b).ensuring(pRef.get.flatMap(_.fold(IO.unit)(t => release(t._1, t._2))))
+          } yield b).ensuring(pRef.get.flatMap(_.fold(IO.unit: IO[Nothing, Any])(t => release(t._1, t._2))))
```
[the remark this answers is not in the document]
Comment (mijicd) [reply core/shared/src/main/scala/scalaz/zio/Promise.scala]: Omitting it causes the following failure:

```
type mismatch;
[error]  found   : scalaz.zio.IO[Nothing,_$28] where type _$28
[error]  required: scalaz.zio.IO[Nothing,Unit]
[error]       } yield b).ensuring(m.get.flatMap(_.fold(unit)(release(_))))
[error]                                                             ^
```

[the remark this answers is not in the document]
Comment (mijicd) [reply core/shared/src/main/scala/scalaz/zio/Promise.scala]: Going with less idiomatic approach via pattern matching does not suffer from the same issue though.

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 3]
```diff
@@ -1118,12 +1118,12 @@ object IO extends Serializable {
...
   final def bracket[E, A, B](
     acquire: IO[E, A]
-  )(release: A => IO[Nothing, Unit])(use: A => IO[E, B]): IO[E, B] =
+  )(release: A => IO[Nothing, _])(use: A => IO[E, B]): IO[E, B] =
     Ref[Option[A]](None).flatMap { m =>
       (for {
         a <- acquire.flatMap(a => m.set(Some(a)).const(a)).uninterruptibly
         b <- use(a)
-      } yield b).ensuring(m.get.flatMap(_.fold(unit)(release(_))))
+      } yield b).ensuring(m.get.flatMap(_.fold(unit: IO[Nothing, Any])(release(_))))
```
[the remark this answers is not in the document]
Comment (mijicd) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: Replaced fold with pattern matching. Less-idiomatic, but the inference works properly.

File: core/shared/src/main/scala/scalaz/zio/Promise.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/Promise.scala]: Does this happen with `IO[Nothing, Any]` too, or just `IO[Nothing, _]`?

**Comment by mijicd** [comment]
@jdegoes it has been resolved by replacing fold with map + getOrElse.

**Comment by jdegoes** [comment]
@mijicd This helps us but doesn't help the users of the library. Can you see if `Any` has any advantage here over wildcards for type inference?

**Comment by mijicd** [comment]
@jdegoes not sure that I understand. From the point of view of problematic usages (i.e. `Option.fold`), both approaches suffer from type inference issues. Replacing `fold` with `map` + `getOrElse` removes the need for adding explicit (and perhaps too broad) type annotations. Reading through the #381 discussion lead me to conclusion that using wildcards is the preferable approach, with the risk of being removed in 3.2.

**Comment by jdegoes** [comment]
> From the point of view of problematic usages (i.e. Option.fold), both approaches suffer from type inference issues.

All right. This is what I was interested in. I am not sure why Scala can't unify them to `IO[Nothing, Any]`.

**Pull request merged**